File: crates/lib/lib-schema/src/models/candidate.rs
use chrono::NaiveDateTime;
use chrono::naive::serde::ts_milliseconds_option::serialize as to_milli_tsopt;
use chrono::naive::serde::ts_milliseconds_option::deserialize as from_milli_tsopt;
use lib_entity::entities::candidate::{Model as DbCandidate, CandidateStatus};
use sea_orm::prelude::*;
use serde::{Deserialize, Serialize};
use serde_json::Value;

use sea_orm::FromQueryResult;

/// 候选人模型
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Candidate {
    /// 唯一标识符
    pub id: i32,
    /// 公司ID
    pub company_id: i32,
    /// 候选人姓名
    pub name: String,
    /// 联系电话
    pub phone: Option<String>,
    /// 电子邮箱
    pub email: Option<String>,
    /// 应聘职位ID
    pub position_id: i32,
    /// 目标部门ID
    pub department_id: i32,
    /// 面试日期
    #[serde(
        serialize_with = "to_milli_tsopt",
        deserialize_with = "from_milli_tsopt"
    )]
    pub interview_date: Option<NaiveDateTime>,
    /// 状态
    pub status: CandidateStatus,
    /// 面试官ID
    pub interviewer_id: Option<i32>,
    /// 面试评价
    #[serde(skip_serializing_if = "Option::is_none")]
    pub evaluation: Option<String>,
    /// 备注
    #[serde(skip_serializing_if = "Option::is_none")]
    pub remark: Option<String>,
    /// 额外字段值（JSON）
    #[serde(skip_serializing_if = "Option::is_none")]
    pub extra_value: Option<Value>,
    /// 额外字段模式ID
    #[serde(skip_serializing_if = "Option::is_none")]
    pub extra_schema_id: Option<i32>,
    /// 创建时间
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(serialize_with = "to_milli_tsopt", deserialize_with = "from_milli_tsopt")]
    pub created_at: Option<NaiveDateTime>,
    /// 更新时间
    #[serde(serialize_with = "to_milli_tsopt", deserialize_with = "from_milli_tsopt")]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub updated_at: Option<NaiveDateTime>,
}

/// 候选人数模型，用于创建和更新
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct InsertCandidate {
    /// 唯一标识符（更新时使用）
    pub id: Option<i32>,
    /// 公司ID
    pub company_id: i32,
    /// 候选人姓名
    pub name: String,
    /// 联系电话
    pub phone: Option<String>,
    /// 电子邮箱
    pub email: Option<String>,
    /// 应聘职位ID
    pub position_id: i32,
    /// 目标部门ID
    pub department_id: i32,
    /// 面试日期
    #[serde(
        serialize_with = "to_milli_tsopt",
        deserialize_with = "from_milli_tsopt"
    )]
    pub interview_date: Option<NaiveDateTime>,
    /// 面试官ID
    pub interviewer_id: Option<i32>,
    /// 额外字段值（JSON）
    pub extra_value: Option<Value>,
    /// 额外字段模式ID
    pub extra_schema_id: Option<i32>,
}

/// 更新候选人状态的请求模型
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UpdateCandidateStatus {
    /// 状态
    pub status: CandidateStatus,
    /// 面试评价
    pub evaluation: Option<String>,
    /// 备注
    pub remark: Option<String>,
}

impl FromQueryResult for Candidate {
    fn from_query_result(res: &QueryResult, pre: &str) -> Result<Self, DbErr> {
        let extra_value: Option<Value> = res
            .try_get::<String>(pre, "extra_value")
            .ok()
            .and_then(|s| serde_json::from_str(&s).ok());

        let status = match res.try_get::<String>(pre, "status")?.as_str() {
            "Pending" => CandidateStatus::Pending,
            "Scheduled" => CandidateStatus::Scheduled,
            "Interviewed" => CandidateStatus::Interviewed,
            "Accepted" => CandidateStatus::Accepted,
            "Rejected" => CandidateStatus::Rejected,
            "Withdrawn" => CandidateStatus::Withdrawn,
            _ => CandidateStatus::Pending, // 默认状态
        };

        Ok(Self {
            id: res.try_get(pre, "id")?,
            company_id: res.try_get(pre, "company_id")?,
            name: res.try_get(pre, "name")?,
            phone: res.try_get(pre, "phone")?,
            email: res.try_get(pre, "email").unwrap_or(None),
            position_id: res.try_get(pre, "position_id")?,
            department_id: res.try_get(pre, "department_id")?,
            interview_date: res.try_get(pre, "interview_date")?,
            status,
            interviewer_id: res.try_get(pre, "interviewer_id")?,
            evaluation: res.try_get(pre, "evaluation").unwrap_or(None),
            remark: res.try_get(pre, "remark").unwrap_or(None),
            extra_value,
            extra_schema_id: res.try_get(pre, "extra_schema_id").ok(),
            created_at: res.try_get(pre, "created_at")?,
            updated_at: res.try_get(pre, "updated_at")?,
        })
    }
}

impl From<DbCandidate> for Candidate {
    fn from(model: DbCandidate) -> Self {
        Self {
            id: model.id,
            company_id: model.company_id,
            name: model.name,
            phone: model.phone,
            email: model.email,
            position_id: model.position_id,
            department_id: model.department_id,
            interview_date: Some(model.interview_date),
            status: model.status,
            interviewer_id: model.interviewer_id,
            evaluation: model.evaluation,
            remark: model.remark,
            extra_value: model.extra_value,
            extra_schema_id: model.extra_schema_id,
            created_at: Some(model.created_at),
            updated_at: Some(model.updated_at),
        }
    }
} 

File: src/lib/api/candidate.ts
import { PageParams, PageResult } from "@/lib/types";
import { serverAPI, ApiResponse } from "./client";
import useSWR from "swr";

export interface CandidateListParams extends PageParams {
  status?: CandidateStatus;
  search?: string;
}

// 候选人状态枚举
export enum CandidateStatus {
  Pending = "Pending", // 待处理
  Scheduled = "Scheduled", // 已安排面试
  Interviewed = "Interviewed", // 已面试
  Accepted = "Accepted", // 已通过
  Rejected = "Rejected", // 已拒绝
  Withdrawn = "Withdrawn", // 已撤回
}

export interface Candidate {
  id: number;
  company_id: number;
  name: string;
  phone: string;
  email?: string | null;
  position_id: number;
  department_id: number;
  interview_date?: number | null;
  status: CandidateStatus;
  interviewer_id: number;
  evaluation?: string | null;
  remark?: string | null;
  extra_value?: any | null;
  extra_schema_id?: number | null;
  created_at?: number | null;
  updated_at?: number | null;
}

export interface InsertCandidate {
  id?: number;
  company_id: number;
  name: string;
  phone?: string | null;
  email?: string | null;
  position_id: number;
  department_id: number;
  interview_date?: string | null;
  interviewer_id?: number | null;
  extra_value?: any | null;
  extra_schema_id?: number | null;
  remark?: string | null;
}

export interface UpdateCandidateStatus {
  status: CandidateStatus;
  evaluation?: string | null;
  remark?: string | null;
}

const API_PREFIX = "candidate";

/**
 * 创建或更新候选人
 * @param data 候选人数据
 */
export async function createOrUpdateCandidate(
  data: InsertCandidate
): Promise<Candidate> {
  let timestamp = null;
  if (data.interview_date) {
    const date = new Date(data.interview_date);
    timestamp = date.getTime();
  }

  const payload = {
    ...data,
    interview_date: timestamp,
  };
  console.log(`data: ${JSON.stringify(payload)}`);
  const response = await serverAPI
    .post(`${API_PREFIX}/insert`, {
      json: payload,
    })
    .json<ApiResponse<Candidate>>();
  if (!response.data) throw new Error("No data returned");
  return response.data;
}

/**
 * 更新候选人状态
 * @param id 候选人ID
 * @param data 状态数据
 */
export async function updateCandidateStatus(
  id: number,
  data: UpdateCandidateStatus
): Promise<Candidate> {
  const response = await serverAPI
    .post(`${API_PREFIX}/${id}/status`, {
      json: data,
    })
    .json<ApiResponse<Candidate>>();
  if (!response.data) throw new Error("No data returned");
  return response.data;
}

/**
 * 获取候选人列表
 * @param companyId 公司ID
 * @param params 分页和筛选参数
 */
export async function getCandidateList(
  companyId: number,
  params: CandidateListParams
): Promise<PageResult<Candidate>> {
  const searchParams: Record<string, string> = {
    page: params.page.toString(),
    limit: params.limit.toString(),
  };

  if (params.status) {
    searchParams.status = params.status;
  }

  if (params.search) {
    searchParams.search = params.search;
  }

  const response = await serverAPI
    .get(`${API_PREFIX}/list/${companyId}`, {
      searchParams,
    })
    .json<ApiResponse<PageResult<Candidate>>>();
  if (!response.data) throw new Error("No data returned");
  return response.data;
}

/**
 * 获取候选人详情
 * @param id 候选人ID
 */
export async function getCandidateById(id: number): Promise<Candidate | null> {
  const response = await serverAPI
    .get(`${API_PREFIX}/get/${id}`)
    .json<ApiResponse<Candidate | null>>();
  return response.data || null;
}

/**
 * 删除候选人
 * @param id 候选人ID
 */
export async function deleteCandidate(id: number): Promise<void> {
  await serverAPI.post(`${API_PREFIX}/delete/${id}`).json<ApiResponse<void>>();
}

/**
 * 使用 SWR 获取候选人列表
 * @param companyId 公司ID
 * @param params 分页和筛选参数
 */
export function useCandidateList(
  companyId: number | undefined,
  params: CandidateListParams
) {
  return useSWR(
    companyId ? ["candidates", "list", companyId, params] : null,
    async () => {
      if (!companyId) return null;
      return await getCandidateList(companyId, params);
    }
  );
}

/**
 * 使用 SWR 获取候选人详情
 * @param id 候选人ID
 */
export function useCandidate(id: number | undefined) {
  return useSWR(id ? ["candidate", id] : null, async () => {
    if (!id) return null;
    return await getCandidateById(id);
  });
}

/**
 * 获取候选人状态的中文描述
 * @param status 状态枚举值
 */
export function getCandidateStatusText(status: CandidateStatus): string {
  const statusMap: Record<CandidateStatus, string> = {
    [CandidateStatus.Pending]: "待处理",
    [CandidateStatus.Scheduled]: "已安排面试",
    [CandidateStatus.Interviewed]: "已面试",
    [CandidateStatus.Accepted]: "已通过",
    [CandidateStatus.Rejected]: "已拒绝",
    [CandidateStatus.Withdrawn]: "已撤回",
  };
  return statusMap[status] || "未知状态";
}

/**
 * 获取所有候选人状态选项
 */
export function getCandidateStatusOptions() {
  return Object.entries(CandidateStatus).map(([, value]) => ({
    label: getCandidateStatusText(value as CandidateStatus),
    value: value,
  }));
}
